### DISCUSSION TITLE
Cannot write to tuning-menu

### DISCUSSION THREAD

**Discussion by gnac** [gotzl/hid-fanatecff#75] [Q&A]
I've installed the hid-fanatecff driver per the instructions. everything went without a hitch.
I plugged in my fanatec DD CSL, and see that it gets registered (see modinfo and dmesg output at bottom).

However, if I try to write to "range" I get an invalid argument error:
```
/sys/module/hid_fanatec/drivers/hid:fanatec/0003:0EB7:0020.0011 $ echo 4 > range
bash: echo: write error: Invalid argument
```
dmesg shows
```
[ 1998.742528] usbhid 2-1:1.0: Cannot write to tuning-menu, not yet read data from device.
```
Which is printed out in hid-ftecff-tuning.c, line 135.
I can however write to 'display' and see the characters on the wheel.

I also tried building the next branch, with similar results.
modinfo on the next modules shows:
```
filename:       /lib/modules/5.14.21-150500.55.52-default/kernel/drivers/hid/hid-fanatec.ko
description:    A driver for the Fanatec CSL Elite
author:         gotzl
license:        GPL
suserelease:    SLE15-SP5
srcversion:     18C66856D563171D5C9E556
alias:          hid:b0003g*v00000EB7p00000011
alias:          hid:b0003g*v00000EB7p00000020
alias:          hid:b0003g*v00000EB7p00000007
alias:          hid:b0003g*v00000EB7p00000006
alias:          hid:b0003g*v00000EB7p00006204
alias:          hid:b0003g*v00000EB7p00000005
alias:          hid:b0003g*v00000EB7p00000E03
alias:          hid:b0003g*v00000EB7p0000183B
alias:          hid:b0003g*v00000EB7p00000004
alias:          hid:b0003g*v00000EB7p00000001
depends:        
retpoline:      Y
name:           hid_fanatec
vermagic:       5.14.21-150500.55.52-default SMP preempt mod_unload modversions 
parm:           init_range:int
parm:           profile:Enable profile debug messages. (int)
parm:           init_load:int
```

Here's the dmesg output when I first plug in the unit:
```
[ 2349.184816] usb 2-1: USB disconnect, device number 15
[ 2352.890516] usb 2-1: new full-speed USB device number 16 using xhci_hcd
[ 2353.040261] usb 2-1: New USB device found, idVendor=0eb7, idProduct=0020, bcdDevice= 0.17
[ 2353.040270] usb 2-1: New USB device strings: Mfr=1, Product=3, SerialNumber=0
[ 2353.040274] usb 2-1: Product: FANATEC Wheel
[ 2353.040276] usb 2-1: Manufacturer: Fanatec
[ 2353.042414] hid_map_usage: 78 callbacks suppressed
[ 2353.042418] Fanatec FANATEC Wheel: Invalid code 768 type 1
[ 2353.042422] Fanatec FANATEC Wheel: Invalid code 769 type 1
[ 2353.042424] Fanatec FANATEC Wheel: Invalid code 770 type 1
[ 2353.042425] Fanatec FANATEC Wheel: Invalid code 771 type 1
[ 2353.042427] Fanatec FANATEC Wheel: Invalid code 772 type 1
[ 2353.042429] Fanatec FANATEC Wheel: Invalid code 773 type 1
[ 2353.042430] Fanatec FANATEC Wheel: Invalid code 774 type 1
[ 2353.042432] Fanatec FANATEC Wheel: Invalid code 775 type 1
[ 2353.042434] Fanatec FANATEC Wheel: Invalid code 776 type 1
[ 2353.042436] Fanatec FANATEC Wheel: Invalid code 777 type 1
[ 2353.042552] input: Fanatec FANATEC Wheel as /devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/0003:0EB7:0020.0012/input/input44
[ 2353.042789] fanatec 0003:0EB7:0020.0012: input,hiddev96,hidraw0: USB HID v1.11 Joystick [Fanatec FANATEC Wheel] on usb-0000:00:14.0-1/input0
[ 2353.105822] fanatec 0003:0EB7:0020.0012: Hires timer: period = 2 ms
```

Thank you.

**Comment by gotzl** [comment]
Hi,

the `range` sysfs file is mapped to the tuning-menu `SEN` value. For writing to this file, first, the current values of the tuning-menu have to be transmitted at least once from the wheel to the PC. This happens for example when you open the tuning-menu on the wheel (I guess), or when you change a value on the wheel directly.

I still have to figure out how I can get the current tuning-menu values from the wheel on initialization ... I'll add this to the 'known issues'. 

Does this sound OK for you?

**Reply by gnac** [reply]
So all I have to do it go into the settings menu on the wheel once?
A known issue would be great. Perhaps a note in the readme in the meantime?
Thanks for all your work BTW.

**Reply by gnac** [reply]
Mucking about with the wheel settings didn't seem to do much re the driver.
However, I switched it to compatibilty mode (Yellow power), and now I can write to range  and I'm getting wheel turn information in oversteer, but pedals are showing as binary, eg one click for each.

**Reply by gotzl** [reply]
Hmm... That's strange. I'm testing with CSL Elite and ClubSport DD, both send tuning-menu values. Maybe I try to get my hands on my brothers CSL DD and see if I can reproduce.

It works in compatibility mode because the driver uses another register on the wheel-base and doesn't interact with the tuning-menu.

In your case, I would stay in native mode and control range setting the tuning-menu on wheel instead of with oversteer...

**Reply by arepavich** [reply]
I'm encountering the same problem with my CSL DD (per duplicate issue #85). While modifying the values from the menu on the wheel itself does appear to modify the behavior of the wheel, it seemingly does not trigger a send and the sysfs values remain uninitialized.

I think this approach works in a pinch for now, though I'd love to see some better compatibility with tools like Oversteer - I'm not well-versed in driver development, but let me know if I can assist in solving this in any way.

**Reply by gotzl** [reply]
Well, this is a discussion and #85 is an issue.. Having both doesn't hurt I guess...
I wonder if either of you could capture the USB traffic (eg using wireshark) when modifying values directly on the wheel. Maybe it sends s.t. I don't expect...

**Reply by arepavich** [reply]
I took a shot at capturing some USB traffic under a few scenarios this evening. There's likely some noise in there with micro-inputs while interacting with the wheel's tuning menu, but I've tried to capture some targeted interactions.
1) Simple USB connection (`csl_dd_on_connect.pcapng`)
2) Opening and closing the tuning menu without making a change (`csl_dd_open_close_tuning_menu.pcapng`)
3) Opening the tuning menu, changing the Sens value, then closing the tuning menu (`csl_dd_update_sens.pcapng`)

[csl_dd_wireshark_captures.zip](https://github.com/user-attachments/files/16104672/csl_dd_wireshark_captures.zip)

**Reply by gotzl** [reply]
Perfect! So it does send packets ;) You can see for yourself by applying the `usb.data_len == 64` filter in wireshark.

There is one difference I spot immediately, which is that the `SLOT` index has to highest bit set (`0x82`). But this only means that reading/writing the `SLOT` sysfs file potentially won't work.
It doesn't explain why it says `.. not yet read data from device`, since there, it checks that slot index is different from `0x00` (which should be true because it apparently is `0x82`).

So, I guess this needs a bit of poking around... Could you add a line just after the `memcpy` at https://github.com/gotzl/hid-fanatecff/blob/124372964d81ab71b4dc54858c746f1fe3621dbe/hid-ftecff.c#L1176C6-L1176C7 reading s.t. like

```
		printk("tuning menu slot: %i\n", drv_data->tuning.ftec_tuning_data[3]);
```

and see if this outputs s.t.? It should end there when a tuning-menu packet is received, so maybe keep wireshark open to see that s.t. was coming...

**Reply by arepavich** [reply]
With that log message in place, it does seem to reach that block on initial connection, as well as whenever a value is updated from the wheel's tuning menu.

```
[   77.979105] usb 5-2: new full-speed USB device number 8 using xhci_hcd
[   78.128456] usb 5-2: New USB device found, idVendor=0eb7, idProduct=0020, bcdDevice= 0.17
[   78.128463] usb 5-2: New USB device strings: Mfr=1, Product=3, SerialNumber=0
[   78.128466] usb 5-2: Product: FANATEC Wheel
[   78.128468] usb 5-2: Manufacturer: Fanatec
[   78.140623] input: Fanatec FANATEC Wheel as /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/input/input28
[   78.140832] fanatec 0003:0EB7:0020.000C: input,hiddev96,hidraw0: USB HID v1.11 Joystick [Fanatec FANATEC Wheel] on usb-0000:0f:00.3-2/input0
[   78.195585] fanatec 0003:0EB7:0020.000C: Hires timer: period = 2 ms
[   78.196387] tuning menu slot: 10
[   78.198386] tuning menu slot: 130
[   99.520661] tuning menu slot: 130
[  105.295631] tuning menu slot: 130
```

**Reply by gotzl** [reply]
But still after getting the log message, reading/writing the sysfs values doesn't work?

**Reply by arepavich** [reply]
Yeah, still getting the same errors when attempting to update the sysfs values. I added an additional print statement to make sure that the following conditional isn't failing, and it appears that we're successfully reaching the `kobject_uevent` call.

**Reply by arepavich** [reply]
I did some further digging into `kobject_uevent` today, and found some advice to use `sudo udevadm monitor` to verify that the udev events are actually being caught. Here's the output on connection of the wheel:
```
> sudo udevadm monitor -u
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing

UDEV  [64413.531619] add      /class/ftec_tuning (class)
UDEV  [64413.534528] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2 (usb)
UDEV  [64413.536409] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0 (usb)
UDEV  [64413.549904] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C (hid)
UDEV  [64413.552795] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/ftec_tuning/0003:0EB7:0020.000C (ftec_tuning)
UDEV  [64413.552911] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM1 (leds)
UDEV  [64413.553392] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM2 (leds)
UDEV  [64413.553827] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM3 (leds)
UDEV  [64413.553923] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM4 (leds)
UDEV  [64413.554419] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM5 (leds)
UDEV  [64413.554674] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM6 (leds)
UDEV  [64413.554737] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM7 (leds)
UDEV  [64413.555353] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM8 (leds)
UDEV  [64413.555452] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/leds/0003:0EB7:0020.000C::RPM9 (leds)
UDEV  [64413.556886] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/hidraw/hidraw9 (hidraw)
UDEV  [64413.637937] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/usbmisc/hiddev2 (usbmisc)
UDEV  [64413.721044] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/input/input30 (input)
UDEV  [64413.968095] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/input/input30/event23 (input)
UDEV  [64413.993779] add      /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/input/input30/js1 (input)
UDEV  [64413.995966] bind     /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C (hid)
UDEV  [64413.997719] bind     /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0 (usb)
UDEV  [64414.001342] bind     /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2 (usb)
UDEV  [64414.002790] change   /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/ftec_tuning/0003:0EB7:0020.000C (ftec_tuning)
UDEV  [64414.009817] change   /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C (hid)
UDEV  [64414.011368] change   /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/ftec_tuning/0003:0EB7:0020.000C (ftec_tuning)
UDEV  [64414.018342] change   /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C (hid)
```

And here's the output on modification of the Sens value from the wheel:
```
> sudo udevadm monitor -u
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing

UDEV  [64790.568241] change   /devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/ftec_tuning/0003:0EB7:0020.000C (ftec_tuning)
```

**Reply by arepavich** [reply]
Now we might be getting somewhere! I started poking around in the directory listed on the change of the Sens value (`sys/devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000C/ftec_tuning/0003:0EB7:0020.000C`), and it looks like all of the values are read in correctly there. Directory structure looks like
```
> ls -al
total 0
drwxr-xr-x 3 root root    0 Jul  6 12:24 .
drwxr-xr-x 3 root root    0 Jul  6 12:24 ..
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 BLI
lrwxrwxrwx 1 root root    0 Jul  6 12:24 device -> ../../../0003:0EB7:0020.000C
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 DPR
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 FEI
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 FF
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 FOR
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 FUL
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 INT
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 NDP
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 NFR
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 NIN
drwxr-xr-x 2 root root    0 Jul  6 12:24 power
--w--w---- 1 root root 4096 Jul  6 12:24 RESET
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 SEN
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 SHO
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 SLOT
-rw-rw-r-- 1 root root 4096 Jul  6 12:24 SPR
lrwxrwxrwx 1 root root    0 Jul  6 12:14 subsystem -> ../../../../../../../../../../class/ftec_tuning
-rw-r--r-- 1 root root 4096 Jul  6 12:24 uevent
```

Values are being initialized and updated correctly in this directory. It doesn't appear that I can overwrite them from the terminal, however; the file permissions do not permit writes as my daily user and, when elevated to root, I can change the value of a given file but the wheel doesn't seem to get the memo.

**Reply by gotzl** [reply]
Ok, so, one thing after another ;) I've just updated the udev-rule in the master branch, this should fix the permissions issue.

Now, which sysfs file did you previously use when you got the `not yet read data from device`?

Can you capture the usb-traffic when writing the sysfs file that allows you read/write (I guess, best use again the `SEN` sysfs file)?

**Reply by arepavich** [reply]
The original sysfs file to which I was trying to write was `/sys/module/hid_fanatec/drivers/hid:fanatec/0003:0EB7:0020.000A/range` (which, predictably, still results in the same error).

After updating to latest master branch and adding my user to the `games` group, I am able to write values into `/sys/devices/pci0000:00/0000:00:08.1/0000:0f:00.3/usb5/5-2/5-2:1.0/0003:0EB7:0020.000A/ftec_tuning/0003:0EB7:0020.000A/SEN` as my user.

Writing to this file does appear to transmit a packet to the wheel, however the wheel still does not appear to update its internal value accordingly.

Here's a sample packet generated after executing `echo "900" > SEN`
[csl_dd_on_write_SEN.zip](https://github.com/user-attachments/files/16117161/csl_dd_on_write_SEN.zip)

**Reply by gotzl** [reply]
> The original sysfs file to which I was trying to write was /sys/module/hid_fanatec/drivers/hid:fanatec/0003:0EB7:0020.000A/range (which, predictably, still results in the same error).

Ah, now I see ;) Just submitted a fix for that.

>Writing to this file does appear to transmit a packet to the wheel, however the wheel still does not appear to update its internal value accordingly.

Still not sure about this one yet, though. Sorry. Will have to think about that. Seems the ClubSport DD behaves different here. Perhaps captures from windows are required here to understand what the CSL DD wants.

**Reply by arepavich** [reply]
Loaded up an old Windows 10 machine today and took a shot at capturing the relevant traffic. Worth noting that I also updated my wheel's firmware along the way, so I'm hoping that doesn't introduce any new variables into the fray here.

Since I don't hope to load up this system frequently, I exported two sets of packets for each action - one filtered on `usb.data_len == 64` and one without that filter. I did my best to get some targeted interactions, but there's definitely some extra traffic on the value update from software as I started the capture just before loading up the Fanatec control panel; it looked like there were a few periodic handshakes/polls occurring while the software was running.
[csl_dd_windows_traffic.zip](https://github.com/user-attachments/files/16120370/csl_dd_windows_traffic.zip)

**Reply by gotzl** [reply]
Ty, I'll have a closer look tomorrow. But one idea would be to just put
```
		drv_data->tuning.ftec_tuning_data[3] = drv_data->tuning.ftec_tuning_data[3]&0xf;
```
just after the memcpy at https://github.com/gotzl/hid-fanatecff/blob/124372964d81ab71b4dc54858c746f1fe3621dbe/hid-ftecff.c#L1176C6-L1176C7

**Reply by gotzl** [reply]
So, it seems the additional bit I had trouble with indicates if the tuning menu is in standard or in advanced mode. In your case it was in standard mode (0x82) while I always had it in advanced mode (0x02). When writing to the device, this bit has to be masked out, similar to what I proposed in the last comment.

I've just pushed a change to the master that does this, and it also allows to read/switch between standard/advanced mode via the `advanced_mode` sysfs file. Could you try this out?

**Reply by arepavich** [reply]
That seems to have done it! Loaded up Oversteer, and the correct values were populated right out of the gate; changes to the settings are now being recognized on the wheel, and (of course) I'm no longer getting errors in the dmesg output.

Also, I had no idea that the advanced mode existed - I should really RTFM :laughing:

Thanks for the quick turnaround!

**Reply by gotzl** [reply]
Great, glad to hear :) Thanks for your help in fixing this!
